### PR TITLE
Convert Secret registry to use update/create strategy, allow filtering by Type

### PR DESCRIPTION
**Pull request by liggitt**
Noticed the secret registry hadn't been updated to make use of the strategy bits in generic etcd. Pulled this one commit out of https://github.com/GoogleCloudPlatform/kubernetes/pull/7101 into it's own PR to get eyes on it separately.
1. Removes a lot of boilerplate around metadata filling
2. Makes generated names work for secrets
3. Makes `type` immutable (it was allowed to change before, though we only had one specified type)
4. Adds `type` as an allowed field filter
5. Changes package structure to match other registries

### CODE REVIEW

**Comment by liggitt** [comment]
@pmorie PTAL

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -1242,6 +1242,29 @@ func ValidateSecret(secret *api.Secret) errs.ValidationErrorList {
 		allErrs = append(allErrs, errs.NewFieldForbidden("data", "Maximum secret size exceeded"))
 	}
 
+	switch secret.Type {
+	case api.SecretTypeOpaque, "":
+		// no-op
+	default:
+		allErrs = append(allErrs, errs.NewFieldInvalid("type", secret.Type, "unknown SecretType"))
```
Comment (smarterclayton) [inline pkg/api/validation/validation.go]: Why can't users define their own secret types?

**Comment by liggitt** [comment]
@nikhiljindal PTAL

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (pmorie) [inline pkg/api/validation/validation.go]: @smarterclayton @liggitt Honestly, I don't feel like types are well-thought out yet.  I have a feeling that docker/rkt pull with secret will help us work through them.  I'm fine with this constraint for now.

File: pkg/api/v1beta2/conversion.go [hunk 2]
```diff
@@ -1564,4 +1564,17 @@ func init() {
 		// If one of the conversion functions is malformed, detect it immediately.
 		panic(err)
 	}
+	err = newer.Scheme.AddFieldLabelConversionFunc("v1beta2", "Secret",
```
Comment (pmorie) [inline pkg/api/v1beta2/conversion.go]: Why are these field label conversions needed?

File: pkg/api/v1beta3/conversion.go [hunk 3]
```diff
@@ -2794,7 +2794,7 @@ func init() {
 		// If one of the conversion functions is malformed, detect it immediately.
 		panic(err)
 	}
-	err = newer.Scheme.AddFieldLabelConversionFunc("v1beta1", "Namespace",
+	err = newer.Scheme.AddFieldLabelConversionFunc("v1beta3", "Namespace",
```
Comment (pmorie) [inline pkg/api/v1beta3/conversion.go]: Seems like a bug you found, ought to be split out into its own PR.

File: pkg/api/v1beta2/conversion.go (hunk 2 above)
Comment (pmorie) [inline pkg/api/v1beta2/conversion.go]: Nevermind, I remember now that you're adding filter on type.

File: pkg/registry/secret/etcd/etcd_test.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+func TestCreate(t *testing.T) {
+	fakeEtcdClient, helper := newHelper(t)
+	storage := NewStorage(helper)
+	test := resttest.New(t, storage, fakeEtcdClient.SetError)
+	secret := validNewSecret("foo")
+	secret.Name = ""
+	secret.GenerateName = "foo-"
+	test.TestCreate(
```
Comment (pmorie) [inline pkg/registry/secret/etcd/etcd_test.go]: Not your problem but I find the signature of TestCreate to be extremely unreadable

File: pkg/registry/secret/strategy.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/fielderrors"
+)
+
+// strategy implements behavior for Secret objects
+type strategy struct {
+	runtime.ObjectTyper
+	api.NameGenerator
+}
+
```
Comment (pmorie) [inline pkg/registry/secret/strategy.go]: Should there be assertions that strategy implements `RESTUpdateStrategy` etc here?

**Comment by pmorie** [comment]
Generally close to LGTM modulo comments.  Probably the biggest thing to work out about this is whether users should be able to create their own types.

File: pkg/registry/secret/strategy.go (hunk 5 above)
Comment (liggitt) [inline pkg/registry/secret/strategy.go]: compiler would catch it when assigned at https://github.com/GoogleCloudPlatform/kubernetes/pull/7419/files#diff-b9cf9c3ecb5fce05f0eb0ee398d2ed0fR61

File: pkg/registry/secret/etcd/etcd_test.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/registry/secret/etcd/etcd_test.go]: > On Apr 28, 2015, at 12:32 PM, Paul Morie <notifications@github.com> wrote:
> 
> In pkg/registry/secret/etcd/etcd_test.go:
> 
> > +			Namespace: api.NamespaceDefault,
> > +		},
> > +		Data: map[string][]byte{
> > +			"test": []byte("data"),
> > +		},
> > +	}
> > +}
> > +
> > +func TestCreate(t *testing.T) {
> > +	fakeEtcdClient, helper := newHelper(t)
> > +	storage := NewStorage(helper)
> > +	test := resttest.New(t, storage, fakeEtcdClient.SetError)
> > +	secret := validNewSecret("foo")
> > +	secret.Name = ""
> > +	secret.GenerateName = "foo-"
> > +	test.TestCreate(
> Not your problem but I find the signature of TestCreate to be extremely unreadable
> 
Noted, people with time to come up with a better one are encouraged to provide pulls to improve it...

File: pkg/registry/secret/strategy.go (hunk 5 above)
Comment (pmorie) [inline pkg/registry/secret/strategy.go]: That's fair, though those assertions are really more for humans reading the
code than the compiler.

File: pkg/api/v1beta3/conversion.go (hunk 3 above)
Comment (liggitt) [inline pkg/api/v1beta3/conversion.go]: so split

**Comment by liggitt** [comment]
comments addressed

File: pkg/registry/secret/etcd/etcd_test.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/rest/resttest"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta3"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools/etcdtest"
+)
+
+func newHelper(t *testing.T) (*tools.FakeEtcdClient, tools.EtcdHelper) {
+	fakeEtcdClient := tools.NewFakeEtcdClient(t)
+	fakeEtcdClient.TestIndex = true
+	helper := tools.NewEtcdHelper(fakeEtcdClient, v1beta3.Codec, etcdtest.PathPrefix())
```
Comment (nikhiljindal) [inline pkg/registry/secret/etcd/etcd_test.go]: You should use testapi.Codec instead of hardcoding v1beta3.Codec. That allows us to test all apiVersions and easily add or remove apiVersions without having to change all tests.

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (pmorie) [inline pkg/api/validation/validation.go]: @liggitt I think we should take this validation out for now and open an issue re: types and whether they can be user defined.

File: pkg/registry/secret/strategy.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+		sa, ok := obj.(*api.Secret)
+		if !ok {
+			return false, fmt.Errorf("not a secret")
+		}
+		fields := SelectableFields(sa)
+		return label.Matches(labels.Set(sa.Labels)) && field.Matches(fields), nil
+	})
+}
+
+// SelectableFields returns a label set that represents the object
```
Comment (nikhiljindal) [inline pkg/registry/secret/strategy.go]: This comment seems misleading.
Shouldnt SelectableFields should return a label set with fields that are allowed in field selector query params? They dont necessarily have to represent the object.

File: pkg/registry/secret/strategy.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+func SelectableFields(obj *api.Secret) labels.Set {
+	return labels.Set{
+		"metadata.name": obj.Name,
```
Comment (nikhiljindal) [inline pkg/registry/secret/strategy.go]: Is metadata.name required here? It is not allowed in field selector.

File: pkg/registry/secret/etcd/etcd_test.go (hunk 6 above)
Comment (pmorie) [inline pkg/registry/secret/etcd/etcd_test.go]: +1 to this, good catch

**Comment by liggitt** [comment]
Update with comments, removed type error, removed metadata.name filtering

File: pkg/api/validation/validation.go [hunk 9]
```diff
@@ -1242,6 +1242,29 @@ func ValidateSecret(secret *api.Secret) errs.ValidationErrorList {
 		allErrs = append(allErrs, errs.NewFieldForbidden("data", "Maximum secret size exceeded"))
 	}
 
+	switch secret.Type {
```
Comment (pmorie) [inline pkg/api/validation/validation.go]: Can't this just be replaced with a TODO to add future validations based on types?

File: pkg/kubectl/resource_printer.go [hunk 10]
```diff
@@ -255,7 +255,7 @@ var eventColumns = []string{"FIRSTSEEN", "LASTSEEN", "COUNT", "NAME", "KIND", "S
 var limitRangeColumns = []string{"NAME"}
 var resourceQuotaColumns = []string{"NAME"}
 var namespaceColumns = []string{"NAME", "LABELS", "STATUS"}
-var secretColumns = []string{"NAME", "DATA"}
+var secretColumns = []string{"NAME", "TYPE", "DATA"}
```
Comment (pmorie) [inline pkg/kubectl/resource_printer.go]: Cannot believe this wasn't in there before.  Thanks for adding.

File: pkg/api/validation/validation.go (hunk 9 above)
Comment (liggitt) [inline pkg/api/validation/validation.go]: cut me a tiny bit of slack... I'm trying to keep the service accounts PR balanced on top of this commit :)

**Comment by nikhiljindal** [comment]
Thanks for the fixes. LGTM.

File: pkg/api/validation/validation.go (hunk 9 above)
Comment (pmorie) [inline pkg/api/validation/validation.go]: ;)

**Comment by pmorie** [comment]
I'll merge this on green.

File: pkg/api/validation/validation.go (hunk 9 above)
Comment (pmorie) [inline pkg/api/validation/validation.go]: Ah, my bad, I had forgotten about the ServiceAccountToken type

**Comment by pmorie** [comment]
It must be raining in the internet @liggitt

**Comment by pmorie** [comment]
I'll merge tomorrow AM during business hours.
